rust2/drivers/dummy/src/lib.rs

```diff
@@ -0,0 +1,890 @@
+    }
+
+    fn get_option_bytes(&self, key: Self::Option) -> Result<Vec<u8>> {
+        get_option_bytes(&self.options, key, "database")
+    }
+
+    fn get_option_double(&self, key: Self::Option) -> Result<f64> {
+        get_option_double(&self.options, key, "database")
+    }
+
+    fn get_option_int(&self, key: Self::Option) -> Result<i64> {
+        get_option_int(&self.options, key, "database")
+    }
+
+    fn get_option_string(&self, key: Self::Option) -> Result<String> {
+        get_option_string(&self.options, key, "database")
+    }
+}
+
+impl Database for DummyDatabase {
+    type ConnectionType = DummyConnection;
+
+    fn new_connection(&mut self) -> Result<Self::ConnectionType> {
+        self.new_connection_with_opts([].into_iter())
+    }
+
+    fn new_connection_with_opts(
+        &mut self,
+        opts: impl Iterator<Item = (<Self::ConnectionType as Optionable>::Option, OptionValue)>,
+    ) -> Result<Self::ConnectionType> {
+        let mut connection = Self::ConnectionType {
+            options: HashMap::new(),
+        };
+        for (key, value) in opts {
+            connection.set_option(key, value)?;
+        }
+        Ok(connection)
+    }
+}
+
+pub struct DummyConnection {
+    options: HashMap<OptionConnection, OptionValue>,
+}
+
+impl Optionable for DummyConnection {
+    type Option = OptionConnection;
+
+    fn set_option(&mut self, key: Self::Option, value: OptionValue) -> Result<()> {
+        set_option(&mut self.options, key, value)
+    }
+
+    fn get_option_bytes(&self, key: Self::Option) -> Result<Vec<u8>> {
+        get_option_bytes(&self.options, key, "connection")
+    }
+
+    fn get_option_double(&self, key: Self::Option) -> Result<f64> {
+        get_option_double(&self.options, key, "connection")
+    }
+
+    fn get_option_int(&self, key: Self::Option) -> Result<i64> {
+        get_option_int(&self.options, key, "connection")
+    }
+
+    fn get_option_string(&self, key: Self::Option) -> Result<String> {
+        get_option_string(&self.options, key, "connection")
+    }
+}
+
+impl Connection for DummyConnection {
+    type StatementType = DummyStatement;
+
+    fn new_statement(&mut self) -> Result<Self::StatementType> {
+        Ok(Self::StatementType {
+            options: HashMap::new(),
+        })
+    }
+
+    // This method is used to test that errors round-trip correctly.
+    fn cancel(&mut self) -> Result<()> {
+        let mut error = Error::with_message_and_status("message", Status::Cancelled);
+        error.vendor_code = constants::ADBC_ERROR_VENDOR_CODE_PRIVATE_DATA;
+        error.sqlstate = [1, 2, 3, 4, 5];
+        error.details = Some(vec![
+            ("key1".into(), b"AAA".into()),
+            ("key2".into(), b"ZZZZZ".into()),
+        ]);
+        Err(error)
+    }
+
+    fn commit(&mut self) -> Result<()> {
+        Ok(())
+    }
+
+    fn get_info(&self, _codes: Option<HashSet<InfoCode>>) -> Result<impl RecordBatchReader> {
```

Comment:
Nit: would be nice if the codes were used in this example.

Reply:
I could try to artifically use the codes but it won't provide much value. If we want to usefully use them it will take a fair amount of work I think.

The goal here is just to test that data round-trip correctly between Rust -> C -> Rust (testing both the driver exporter & manager).